torch/csrc/jit/passes/memory_planning/memory_observer.cpp

```diff
@@ -0,0 +1,165 @@
+//
+// Created by Maksim Levental on 9/11/21.
+//
+
+#include <iostream>
+#include <utility>
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/passes/memory_planning/memory_observer.h>
+#include <torch/csrc/jit/runtime/interpreter.h>
+
+namespace torch {
+namespace jit {
+
+MemoryObserverThreadLocalState* getMemoryObserverTLSState() {
+  const auto& state =
+      c10::ThreadLocalDebugInfo::get(c10::DebugInfoKind::PROFILER_STATE);
+  return static_cast<MemoryObserverThreadLocalState*>(state);
+}
+
+struct MemoryObserverContext : public at::ObserverContext {
+  MemoryObserverContext(
+      std::vector<intptr_t> input_ival_addrs,
+      std::vector<std::string> input_val_names,
+      std::vector<std::string> output_val_names,
+      std::string fn_name,
+      int64_t start_time)
+      : input_ival_addrs(std::move(input_ival_addrs)),
+        input_val_names(std::move(input_val_names)),
+        output_val_names(std::move(output_val_names)),
+        fn_name(std::move(fn_name)),
+        start_time(start_time) {}
+  std::vector<intptr_t> input_ival_addrs;
+  std::vector<std::string> input_val_names;
+  std::vector<std::string> output_val_names;
+  std::string fn_name;
+  int64_t start_time;
+};
+
+void pushObserverCallbacks() {
```

Comment:
what are we trying to achieve here ?

Reply:
spy on allocations! and match them up with inputs/outputs to nodes that otherwise only have profile data